src/persist-client/src/operators/shard_source.rs

```diff
@@ -0,0 +1,506 @@
+            // contain only times greater or equal to `until`, which means they can be dropped
+            // in their entirety. The current batch must be emitted, but we can stop afterwards.
+            if PartialOrder::less_equal(&until, &progress) {
+                done = true;
+            }
+
+            let parts_size_bytes = parts.iter().map(|x| x.encoded_size_bytes()).sum();
+            yield (parts, progress, parts_size_bytes);
+        }
+
+        // Rather than simply end, we spawn a task that can continue to return
+        // leases. This keeps the `ReadHandle` alive until we have confirmed
+        // each lease as read.
+        //
+        // TODO: Replace the channel, this task, and the shard_source_tokens
+        // operator with a timely edge between the shard_source_fetch output and
+        // an input exchanging to the chosen worker for shard_source_descs.
+        mz_ore::task::spawn(|| "LeaseReturner", async move {
+            while let Some(leased_part) = consumed_part_rx.recv().await {
+                subscription
+                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
+            }
+        });
+    });
+
+    // TODO: Now that this operator is async AND the pinned stream does not
+    // participate in a select! call, which means its methods never get
+    // cancelled, we can get rid of the whole async_stream! macro machinery and
+    // just get the value directly with a function call.
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("shard_source_descs({})", name), scope.clone());
+    // Disconnect the flow_control_input from the output capabilities of the
+    // operator. We could leave it connected without risking deadlock so long as
+    // there is a non zero summary on the feedback edge. But it may be less
+    // efficient because the pipeline will be moving in increments of SUMMARY
+    // even though we have potentially dumped a lot more data in the pipeline
+    // because of batch boundaries. Leaving it unconnected means the pipeline
+    // will be able to retire bigger chunks of work.
+    let (mut descs_output, descs_stream) = builder.new_output();
+    let mut flow_control_input = builder.new_input_connection(
+        flow_control_input,
+        Pipeline,
+        // Disconnect the flow_control_input from the output capabilities of the
+        // operator. We could leave it connected without risking deadlock so
+        // long as there is a non zero summary on the feedback edge. But it may
+        // be less efficient because the pipeline will be moving in increments
+        // of SUMMARY even though we have potentially dumped a lot more data in
+        // the pipeline because of batch boundaries. Leaving it unconnected
+        // means the pipeline will be able to retire bigger chunks of work.
```

Comment:
Excuse the drive-by comment, but I assume you didn't want to include this comment twice :)

Reply:
done!